File: harness/legacy_governance_read_import_contract/memory.py
"""Strict, loss-preserving parser for supplied legacy memory JSONL bytes."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from typing import Any

from .canonical import ContractError, validate_text
from .constants import (
    MAX_ARRAY_ITEMS, MAX_CONFIDENCE_LEXEME_BYTES, MAX_DEPTH, MAX_FIELDS,
    MAX_MEMORY_ENTRIES, MAX_MEMORY_LINE_BYTES, MAX_SOURCE_REF_BYTES,
)

FIELDS = {
    "_format", "kind", "topic", "detail", "iteration", "source", "confidence",
    "supersedes", "created_at_unix",
}
REQUIRED = {"kind", "topic", "detail", "iteration", "created_at_unix"}
INTEGER_RE = re.compile(r"-?(?:0|[1-9][0-9]*)")
DECIMAL_RE = re.compile(
    r"(-?)(0|[1-9][0-9]*)(?:\.([0-9]+))?(?:[eE]([+-]?[0-9]+))?"
)


@dataclass(frozen=True)
class NumberLexeme:
    raw: str


def _number(raw: str) -> NumberLexeme:
    return NumberLexeme(raw)


def _constant(raw: str) -> None:
    raise ContractError(f"non-finite JSON number {raw!r} is forbidden")


def _pairs(values: list[tuple[str, Any]]) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for key, value in values:
        if key in result:
            raise ContractError(f"duplicate legacy memory key {key!r}")
        result[key] = value
    return result


def _bounded_shape(value: Any, depth: int = 1) -> None:
    if depth > MAX_DEPTH:
        raise ContractError(f"legacy memory JSON depth exceeds {MAX_DEPTH}")
    if isinstance(value, dict):
        if len(value) > MAX_FIELDS:
            raise ContractError(f"legacy memory object fields exceed {MAX_FIELDS}")
        for child in value.values():
            _bounded_shape(child, depth + 1)
    elif isinstance(value, list):
        if len(value) > MAX_ARRAY_ITEMS:
            raise ContractError(f"legacy memory array items exceed {MAX_ARRAY_ITEMS}")
        for child in value:
            _bounded_shape(child, depth + 1)


def _integer(value: Any, label: str) -> int:
    if not isinstance(value, NumberLexeme) or INTEGER_RE.fullmatch(value.raw) is None:
        raise ContractError(f"{label} must be a JSON integer")
    digits = value.raw[1:] if value.raw.startswith("-") else value.raw
    if len(digits) > 19:
        raise ContractError(f"{label} is outside signed int64")
    try:
        integer = int(value.raw)
    except ValueError as error:
        raise ContractError(f"{label} is outside signed int64") from error
    if not -(2**63) <= integer <= 2**63 - 1:
        raise ContractError(f"{label} is outside signed int64")
    return integer


def _optional_text(entry: dict[str, Any], field: str) -> str | None:
    if field not in entry:
        return None
    return validate_text(entry[field], f"legacy memory {field}", MAX_SOURCE_REF_BYTES)


def _confidence(entry: dict[str, Any]) -> dict[str, str | None]:
    if "confidence" not in entry:
        return {"presence": "omitted", "raw_number_lexeme": None}
    value = entry["confidence"]
    if not isinstance(value, NumberLexeme):
        raise ContractError("legacy memory confidence must be a JSON number")
    if not 1 <= len(value.raw.encode("ascii")) <= MAX_CONFIDENCE_LEXEME_BYTES:
        raise ContractError("legacy memory confidence number lexeme is too long")
    if not _confidence_in_range(value.raw):
        raise ContractError("legacy memory confidence must be in decimal range 0..1")
    return {"presence": "explicit", "raw_number_lexeme": value.raw}


def _confidence_in_range(raw: str) -> bool:
    match = DECIMAL_RE.fullmatch(raw)
    if match is None:
        return False
    sign, integer, fraction, exponent = match.groups()
    fraction = fraction or ""
    digits = (integer + fraction).lstrip("0")
    if not digits:
        return True
    if sign:
        return False
    position = len(digits) - len(fraction) + int(exponent or "0")
    if position < 1:
        return True
    if position > 1 or digits[0] != "1":
        return False
    return not digits[1:].strip("0")


def parse_line(raw: bytes, ordinal: int) -> dict[str, Any]:
    if not 1 <= len(raw) <= MAX_MEMORY_LINE_BYTES:
        raise ContractError(f"legacy memory line {ordinal} exceeds its byte bound")
    try:
        entry = json.loads(raw.decode("utf-8"), object_pairs_hook=_pairs,
                           parse_int=_number, parse_float=_number,
                           parse_constant=_constant)
    except (UnicodeDecodeError, json.JSONDecodeError, RecursionError) as error:
        raise ContractError(f"legacy memory line {ordinal} is invalid JSON: {error}") from error
    if not isinstance(entry, dict):
        raise ContractError(f"legacy memory line {ordinal} must be one JSON object")
    _bounded_shape(entry)
    if unknown := set(entry) - FIELDS:
        raise ContractError(f"legacy memory line {ordinal} has unknown fields {sorted(unknown)!r}")
    if missing := REQUIRED - set(entry):
        raise ContractError(f"legacy memory line {ordinal} lacks fields {sorted(missing)!r}")
    legacy_format = entry.get("_format")
    if "_format" in entry and legacy_format != "forgeos.memory.v1":
        raise ContractError("legacy memory _format must be omitted or forgeos.memory.v1")
    kind = entry["kind"]
    if not isinstance(kind, str) or kind not in {"decision", "gap", "lesson"}:
        raise ContractError("legacy memory kind is outside the frozen vocabulary")
    return {
        "confidence": _confidence(entry),
        "created_at_unix": _integer(entry["created_at_unix"], "created_at_unix"),
        "declared_kind": kind,
        "declared_source": _optional_text(entry, "source"),
        "declared_supersedes": _optional_text(entry, "supersedes"),
        "declared_topic": validate_text(entry["topic"], "legacy memory topic",
                                        MAX_SOURCE_REF_BYTES),
        "detail": validate_text(entry["detail"], "legacy memory detail",
                                MAX_MEMORY_LINE_BYTES),
        "iteration": _integer(entry["iteration"], "iteration"),
        "legacy_format": legacy_format,
    }


def parse_jsonl(raw: bytes) -> list[tuple[int, bytes, dict[str, Any]]]:
    lines = raw[:-1].split(b"\n")
    if len(lines) > MAX_MEMORY_ENTRIES:
        raise ContractError(f"legacy memory entry count exceeds {MAX_MEMORY_ENTRIES}")
    if any(not line for line in lines):
        raise ContractError("legacy memory JSONL must not contain blank lines")
    return [(index, line, parse_line(line, index))
            for index, line in enumerate(lines, start=1)]

File: forge-core/internal/legacygovernanceimportcontract/memory.go
package legacygovernanceimportcontract

import (
	"bytes"
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

var (
	memoryFields = map[string]bool{
		"_format": true, "kind": true, "topic": true, "detail": true,
		"iteration": true, "source": true, "confidence": true,
		"supersedes": true, "created_at_unix": true,
	}
	memoryRequired = []string{"kind", "topic", "detail", "iteration", "created_at_unix"}
	integerPattern = regexp.MustCompile(`^-?(0|[1-9][0-9]*)$`)
	decimalPattern = regexp.MustCompile(
		`^(-?)(0|[1-9][0-9]*)(?:\.([0-9]+))?(?:[eE]([+-]?[0-9]+))?$`,
	)
)

func parseMemoryJSONL(raw []byte) ([]map[string]any, [][]byte, error) {
	lines := bytes.Split(raw[:len(raw)-1], []byte{'\n'})
	if len(lines) > maxMemoryEntries {
		return nil, nil, fmt.Errorf("memory entry count exceeds %d", maxMemoryEntries)
	}
	entries := make([]map[string]any, 0, len(lines))
	for index, line := range lines {
		if len(line) == 0 || len(line) > maxMemoryLineBytes {
			return nil, nil, fmt.Errorf("memory line %d is blank or exceeds bound", index+1)
		}
		entry, err := parseMemoryLine(line, index+1)
		if err != nil {
			return nil, nil, err
		}
		entries = append(entries, entry)
	}
	return entries, lines, nil
}

func parseMemoryLine(raw []byte, ordinal int) (map[string]any, error) {
	value, err := parseStrictJSON(raw, maxMemoryLineBytes, true)
	if err != nil {
		return nil, fmt.Errorf("memory line %d: %w", ordinal, err)
	}
	entry, ok := value.(map[string]any)
	if !ok {
		return nil, fmt.Errorf("memory line %d must be one JSON object", ordinal)
	}
	for field := range entry {
		if !memoryFields[field] {
			return nil, fmt.Errorf("memory line %d has unknown field %q", ordinal, field)
		}
	}
	for _, field := range memoryRequired {
		if _, exists := entry[field]; !exists {
			return nil, fmt.Errorf("memory line %d lacks field %q", ordinal, field)
		}
	}
	if format, present := entry["_format"]; present && format != "forgeos.memory.v1" {
		return nil, fmt.Errorf("memory _format must be omitted or forgeos.memory.v1")
	}
	kind, ok := entry["kind"].(string)
	if !ok || (kind != "decision" && kind != "gap" && kind != "lesson") {
		return nil, fmt.Errorf("memory kind is outside frozen vocabulary")
	}
	return projectMemoryFields(entry, kind)
}

func projectMemoryFields(entry map[string]any, kind string) (map[string]any, error) {
	topic, err := requiredMemoryText(entry, "topic", maxSourceRefBytes)
	if err != nil {
		return nil, err
	}
	detail, err := requiredMemoryText(entry, "detail", maxMemoryLineBytes)
	if err != nil {
		return nil, err
	}
	iteration, err := memoryInteger(entry["iteration"], "iteration")
	if err != nil {
		return nil, err
	}
	created, err := memoryInteger(entry["created_at_unix"], "created_at_unix")
	if err != nil {
		return nil, err
	}
	confidence, err := memoryConfidence(entry)
	if err != nil {
		return nil, err
	}
	source, err := optionalMemoryText(entry, "source")
	if err != nil {
		return nil, err
	}
	supersedes, err := optionalMemoryText(entry, "supersedes")
	if err != nil {
		return nil, err
	}
	return map[string]any{
		"confidence": confidence, "created_at_unix": created, "declared_kind": kind,
		"declared_source": source, "declared_supersedes": supersedes,
		"declared_topic": topic, "detail": detail, "iteration": iteration,
		"legacy_format": entry["_format"],
	}, nil
}

func requiredMemoryText(entry map[string]any, field string, maximum int) (string, error) {
	value, ok := entry[field].(string)
	if !ok || validateWireString(value, maximum, true) != nil {
		return "", fmt.Errorf("memory %s must be bounded nonempty text", field)
	}
	return value, nil
}

func optionalMemoryText(entry map[string]any, field string) (any, error) {
	value, exists := entry[field]
	if !exists {
		return nil, nil
	}
	text, ok := value.(string)
	if !ok || validateWireString(text, maxSourceRefBytes, true) != nil {
		return nil, fmt.Errorf("memory %s must be bounded nonempty text", field)
	}
	return text, nil
}

func memoryInteger(value any, label string) (int64, error) {
	lexeme, ok := value.(numberLexeme)
	if !ok || !integerPattern.MatchString(string(lexeme)) {
		return 0, fmt.Errorf("memory %s must be a JSON integer", label)
	}
	integer, err := strconv.ParseInt(string(lexeme), 10, 64)
	if err != nil {
		return 0, fmt.Errorf("memory %s is outside signed int64", label)
	}
	return integer, nil
}

func memoryConfidence(entry map[string]any) (map[string]any, error) {
	value, exists := entry["confidence"]
	if !exists {
		return map[string]any{"presence": "omitted", "raw_number_lexeme": nil}, nil
	}
	lexeme, ok := value.(numberLexeme)
	if !ok || len(lexeme) == 0 || len(lexeme) > maxConfidenceLexeme ||
		!confidenceInRange(string(lexeme)) {
		return nil, fmt.Errorf("memory confidence must be a decimal number in range 0..1")
	}
	return map[string]any{"presence": "explicit", "raw_number_lexeme": string(lexeme)}, nil
}

func confidenceInRange(raw string) bool {
	parts := decimalPattern.FindStringSubmatch(raw)
	if parts == nil {
		return false
	}
	digits := strings.TrimLeft(parts[2]+parts[3], "0")
	if digits == "" {
		return true
	}
	if parts[1] == "-" {
		return false
	}
	exponent, overflow := parseExponent(parts[4])
	if overflow > 0 {
		return false
	}
	if overflow < 0 {
		return true
	}
	threshold := int64(1 - (len(digits) - len(parts[3])))
	if exponent < threshold {
		return true
	}
	if exponent > threshold || digits[0] != '1' {
		return false
	}
	return strings.Trim(digits[1:], "0") == ""
}

func parseExponent(raw string) (int64, int) {
	if raw == "" {
		return 0, 0
	}
	value, err := strconv.ParseInt(raw, 10, 64)
	if err == nil {
		return value, 0
	}
	if raw[0] == '-' {
		return 0, -1
	}
	return 0, 1
}
